### PR TITLE
Recaptchav3

### PR DESCRIPTION
**Pull request by Freshrojek**
Hey guys, here's the recpatchav3 if you could have a look at it. To go to the register page you have to go /memberships/logout. The next part after registration doesn't work (39 has been done now but wasn't at the time), so basically if you don't get a 403 error it thinks you're human.

### CODE REVIEW

**Comment by CarwynNelson** [comment]
I'll try and take a look at this some time on Friday

**Comment by jamesgeddes** [comment]
Great work @Freshrojek :smile: I will take a look later today

File: memberships/templates/memberships/register.html [hunk 1]
```diff
@@ -2,6 +2,21 @@
 
 {% block content %}
   <h2>Member Registration</h2>
+  <script>
+/*Execute recaptchaV3 function*/
+grecaptcha.ready(function() {
+  grecaptcha.execute('6LfGd9oZAAAAAEx2_gzbS1cR8X0_IEmPNFG8A9eN', {action: "/register/"}).then(function(token) {
```
Comment (CarwynNelson) [inline memberships/templates/memberships/register.html:8]: Could we pull these keys into environment config and then only add this code to the page if those keys are available? That gives us two things:
1. During deployment we can securely add the API keys to the machine without them being committed to the git repoi
2. Developer machines will only have recaptcha if they want them to. That is especially helpful if you are running automated tests that will probably be rejected by recaptcha

There are some downsides to this of course, mainly:
1. Developers are less likely to implicitly test recaptcha when doing development as it is disabled
2. There will have to be extra code in the view to account for the recaptcha key not being sent

File: web/settings.py [hunk 2]
```diff
@@ -23,6 +23,11 @@
 # Quick-start development settings - unsuitable for production
 # See https://docs.djangoproject.com/en/3.0/howto/deployment/checklist/
 
+# recaptchaV3 keys
+#FIXME
+RECAPTCHA_SITE_KEY = '6LfGd9oZAAAAAEx2_gzbS1cR8X0_IEmPNFG8A9eN'
```
Comment (CarwynNelson) [inline web/settings.py:28]: As per the above comment, could we set these from environment variables? :)

File: memberships/views.py [hunk 3]
```diff
@@ -5,30 +5,55 @@
...
     if not form.is_valid():
         return render(request, "memberships/register.html", {"form": form})
-
     if not form.cleaned_data["preferred_name"]:
         form.cleaned_data["preferred_name"] = form.cleaned_data["full_name"]
 
+    recaptchaV3_response = request.POST.get('recaptchaV3-response')
+    success = validate_recaptcha(recaptchaV3_response)
+    if success == 'fail':
+        return HttpResponse('Invalid reCAPTCHA response. Please try again.', status=403)
```
Comment (CarwynNelson) [inline memberships/views.py:55]: I actually got a failed recaptcha response once (but without the checkbox) so I know this works, which is great :)

It does make me think that it's possible for normal humans to get this response though, so maybe in a future ticket we can clean this up and redirect back to the register page with a message that explains that google thinks they are a robot?

**Review by CarwynNelson — COMMENTED** [review]
Looks great so far, just a few small comments :)

I'll try and get your branch re-based on to the main one so you don't have to sort that out. Before you make any changes you might have to do a `git pull`.

Let me know if you have any issues.

**Comment by CarwynNelson** [comment]
Apologies if the formatting is a little different after the rebase. It was quite out of date with the main branch so it was a little tricky to sort out and I had to do some reformatting :)

File: memberships/templates/memberships/register.html (hunk 1 above)
Comment (CarwynNelson) [reply memberships/templates/memberships/register.html]: Another good reason to do this is that this causes the tests to fail as they do not submit a valid recaptcha :)

I suppose we could mock out google, but I tend to prefer avoiding mocks where possible.

Comment (Freshrojek) [reply memberships/templates/memberships/register.html]: Thanks! Will get working on it

File: memberships/views.py (hunk 3 above)
Comment (Freshrojek) [reply memberships/views.py:59]: Yeah that makes sense to me, explaining they didn't fill the form out correctly rather than being a robot?

**Comment by Freshrojek** [comment]
> 
> 
> Apologies if the formatting is a little different after the rebase. It was quite out of date with the main branch so it was a little tricky to sort out and I had to do some reformatting :)

When trying git pull I get this 'Automatic merge failed; fix conflicts and then commit the result.' In register.html and views.py. Is there something I could do to merge it?

File: memberships/views.py (hunk 3 above)
Comment (CarwynNelson) [reply memberships/views.py:59]: Yeah I think that makes sense to me. Or even say that google thinks they are a robot rather than we do.

**Comment by CarwynNelson** [comment]
> > Apologies if the formatting is a little different after the rebase. It was quite out of date with the main branch so it was a little tricky to sort out and I had to do some reformatting :)
> 
> When trying git pull I get this 'Automatic merge failed; fix conflicts and then commit the result.' In register.html and views.py. Is there something I could do to merge it?

That's probably because of the re-base I did. Try the following:
```sh
git fetch --all
git reset --hard origin/recaptchav3
```

**Comment by jamesgeddes** [comment]
Currently not passing the tests, so that needs some attention. Otherwise, lgtm, I'm happy if @CarwynNelson is.

**Comment by jamesgeddes** [comment]
@Freshrojek and @CarwynNelson What else is needed to get this one over the line? Do the tests need updating?

**Comment by CarwynNelson** [comment]
> @Freshrojek and @CarwynNelson What else is needed to get this one over the line? Do the tests need updating?

I think the best way around the failing tests is to allow this feature to be disabled so the tests don't have to work around it at all.

Another positive of this, as mentioned above, is that it makes local setup and running quite a bit easier.

**Comment by CarwynNelson** [comment]
To get this moving I have applied the fixes I recommended. I hope this frees @Freshrojek up to work on something new and interesting :)

PS: I hope that is okay with you Jack, sorry if it was not.

**Review by CarwynNelson — APPROVED** [review]

**Review by CarwynNelson — APPROVED** [review]

**Comment by jamesgeddes** [comment]
SGMT

**Review by Freshrojek — COMMENTED** [review]
Thanks Carwyn for doing this!

**Pull request merged**